File: solanity_firebase_flutter/lib/models/post.model.dart
import 'package:cloud_firestore/cloud_firestore.dart';

class PostModel {
  final String id;
  final String userId;
  final String caption;
  final String mediaUrl;
  final String mediaType;
  final int likesCount;
  final int commentsCount;
  final int sharesCount;
  final Map<String, int> reactionsCount;
  final DateTime? createdAt;

  PostModel({
    required this.id,
    required this.userId,
    required this.caption,
    required this.mediaUrl,
    required this.mediaType,
    this.likesCount = 0,
    this.commentsCount = 0,
    this.sharesCount = 0,
    this.reactionsCount =
        const {}, //The default value of an optional parameter must be constant
    this.createdAt,
  });

  factory PostModel.fromDoc(DocumentSnapshot<Map<String, dynamic>> doc) {
    final data = doc.data()!;
    return PostModel(
      id: doc.id,
      userId: data['userId'],
      caption: data['caption'] ?? '',
      mediaUrl: data['mediaUrl'],
      mediaType: data['mediaType'] ?? 'image',
      likesCount: data['likesCount'] ?? 0,
      commentsCount: data['commentsCount'] ?? 0,
      sharesCount: data['sharesCount'] ?? 0,
      reactionsCount: Map<String, int>.from(data['reactionsCount'] ?? {}),
      createdAt: (data['createdAt'] as Timestamp?)?.toDate(),
    );
  }
}

File: solanity_mern/server/src/models/post.model.ts
import { Schema, model, Document, Types } from 'mongoose';

export interface IPost extends Document {
  _id: Types.ObjectId;
  userId: Types.ObjectId;
  caption: string;
  mediaUrl: string;
  mediaType: 'image' | 'video' | 'pdf';
  likesCount: number;
  commentsCount: number;
  sharesCount: number;
  reactionsCount: Map<string, number>;
  createdAt: Date;
}

const postSchema = new Schema<IPost>(
  {
    userId: { type: Schema.Types.ObjectId, ref: 'User', required: true, index: true },
    caption: { type: String, default: '' },
    mediaUrl: { type: String, required: true },
    mediaType: { type: String, enum: ['image', 'video', 'pdf'], default: 'image' },
    likesCount: { type: Number, default: 0 },
    commentsCount: { type: Number, default: 0 },
    sharesCount: { type: Number, default: 0 },
    reactionsCount: { type: Map, of: Number, default: {} },
  },
  { timestamps: { createdAt: true, updatedAt: false } }
);

postSchema.index({ createdAt: -1 });

export const Post = model<IPost>('Post', postSchema);
